lib/identity_cache/query_api.rb

```diff
@@ -398,18 +404,30 @@ def prefetch_one_association(association, records)
     private
 
     def fetch_recursively_cached_association(ivar_name, association_name) # :nodoc:
-      if IdentityCache.should_use_cache?
-        ivar_full_name = :"@#{ivar_name}"
+      if IdentityCache.fetch_returns_relation
+        if IdentityCache.should_use_cache?
+          ivar_full_name = :"@#{ivar_name}"
+          assoc = if instance_variable_defined?(ivar_full_name)
+            instance_variable_get(ivar_full_name)
+          else
+            instance_variable_set(ivar_full_name, send(association_name))
+          end
 
-        assoc = if instance_variable_defined?(ivar_full_name)
-          instance_variable_get(ivar_full_name)
+          assoc.is_a?(ActiveRecord::Associations::CollectionAssociation) ? assoc.reader : assoc
         else
-          instance_variable_set(ivar_full_name, send(association_name))
+          send(association_name.to_sym)
         end
-
-        assoc.is_a?(ActiveRecord::Associations::CollectionAssociation) ? assoc.reader : assoc
       else
-        send(association_name.to_sym)
+        if IdentityCache.should_use_cache?
+          ivar_full_name = :"@#{ivar_name}"
+          if instance_variable_defined?(ivar_full_name)
+            instance_variable_get(ivar_full_name)
+          else
+            instance_variable_set(ivar_full_name, association(association_name).load_target)
```

Comment:
I think it will be easier to reason about the code if we always store the samr thing in this instance variable and just call to_a on the result if IdentityCache.fetch_returns_relation returns false